---
title: "Lymph Node Single-Cell Cluster Annotation"
author: "Eileen Owens"
date: "2024-11-04"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(dev = "png",
                      dpi = 300,
                      echo = TRUE,
                      cache = TRUE,
                      warning = FALSE,
                      message = FALSE,
                      keep_md = TRUE)
```

# Introduction
The purpose of this script is to annotate single-cell RNA-seq clusters following filtering, normalization, and clustering of the data with Seurat. This script has been adapted from the Seurat (https://satijalab.org/seurat/) and SingleR (https://bioconductor.org/packages/devel/bioc/vignettes/SingleR/inst/doc/SingleR.html) documentation.

# Software packages
```{r software-load, results="hide", message=FALSE}
library(Seurat)
library(tidyverse)
library(patchwork)
library(pheatmap)
library(RColorBrewer)
library(data.table)
library(knitr)
library(clusterProfiler)
```

# Data
## Working directory
```{r setwd}
setwd("C:/Users/edlarsen/Documents/240828_scRNAseq/Cluster_Annotation")
```

## Filtered, normalized, and clustered Seurat object(s)
```{r data-import}
ln_seurat <- readRDS(file = "../Normalization_and_Clustering/LN_NormalizedAndClustered.RData")
```

# Differentially Expressed Features
By default, identifies all positive and negative markers of a single cluster compared to all other cells.

```{r FindAllMarkers-LN, message=FALSE, fig.width = 8, fig.height = 6}

ln_seurat <- FindClusters(ln_seurat, resolution = 0.1)
ln_seurat <- RunUMAP(ln_seurat, dims = 1:10)
ln.markers <- FindAllMarkers(ln_seurat, only.pos=TRUE)

# export
write.csv(ln.markers, file="LN_DEfeaturesByUMAPcluster_res0.1.csv")

# umap
DimPlot(ln_seurat,
        reduction = "umap",
                   label = TRUE,
                   label.size = 6) + 
  plot_annotation(title = "Canine Lymph Node, Resolution: 0.1", theme = theme(plot.title = element_text(hjust = 0.5, size = 20)))
```

## Top cluster markers
```{r topClusterMarkers, fig.width = 18, fig.height = 8}
key.genes <- ln.markers[!grepl("ENSCAFG", row.names(ln.markers)),]
key.genes.sortedByPval <- key.genes[order(key.genes$p_val),]
features <- key.genes.sortedByPval %>% group_by(cluster) %>% do(head(., n=5))
features <- as.data.frame(features[!duplicated(features$gene),]) # prioritize unique clusters

DotPlot(ln_seurat,
        assay = "RNA",
        features = rev(features$gene),
        scale = TRUE) + 
  geom_point(aes(size=pct.exp), shape = 21, colour="black", stroke=0.5) +
  labs(y = "Cluster") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```

# Feature plots
Some of these markers were derived from a single-cell atlas of circulating canine leukocytes (https://github.com/dyammons/Canine_Leukocyte_scRNA).
```{r FeaturePlotLN-CD4markers, fig.width = 14, fig.height = 12}
FeaturePlot(ln_seurat, 
            reduction = "umap",
            ncol = 3,
            features = c("CD3E", "CD5", "CD4", "FOXP3", "IL2RA", "GATA3", "TBX21", "RORC")) +
  plot_annotation(title = "Canine Lymph Node: Expression of Canonical CD4 T-cell Markers", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```

```{r FeaturePlotLN-CD8andNKmarkers, fig.height = 8, fig.width = 14}
FeaturePlot(ln_seurat, 
            reduction = "umap",
            features = c("GNLY", "NKG7", "NCAM1", "KLRK1", "CD8A", "CD3E", "CD5"),
            ncol = 3) +
  plot_annotation(title = "Canine Lymph Node: Expression of Canonical CD8 & NK Cell Markers", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```

```{r FeaturePlotLN-Bcellmarkers, fig.height = 8, fig.width = 10}
FeaturePlot(ln_seurat, 
            reduction = "umap",
            features = c("MS4A1", "CD19", "PAX5", "IRF4")) +
  plot_annotation(title = "Canine Lymph Node: Expression of Canonical B-cell Markers", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```

```{r FeaturePlotLN-MonoMacDCmarkers, fig.height = 8, fig.width = 14}
FeaturePlot(ln_seurat, 
            reduction = "umap",
            ncol = 3,
            features = c("LYZ", "FCER1A", "ITGAM", "CD163", "CD68")) +
  plot_annotation(title = "Canine Lymph Node: Expression of Canonical Monocyte/Macrophage/DC Markers", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```

```{r FeaturePlotLN-GranulocyteMarkers, fig.height = 8, fig.width = 14}
FeaturePlot(ln_seurat, 
            reduction = "umap",
            features = c("FCER1A", "MMP9", "CCR5", "CXCR2", "EPX", "MMP8", "S100A12", "SERPINA1"),
            ncol = 3) +
  plot_annotation(title = "Canine Lymph Node: Expression of Canonical Granulocyte Markers", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```
Note: Granulocytes are difficult to capture with single-cell RNA-seq experiments given their relatively low RNA content and relatively high levels of RNAses and other inhibitory compounds, resulting in fewer detectable transcripts.

# Gene Set Enrichment Analysis
Comparison with gene signatures from a human immune cell atlas and canine circulating leukocyte atlas (https://cells.ucsc.edu/?ds=pan-immune+global; https://www.science.org/doi/10.1126/science.aay3224).

```{r GSEA, warning=FALSE, message=FALSE}
# significant results only
sig.ln.markers <- subset(ln.markers, p_val_adj < 0.05)

# loop through each cluster and create a ranked gene list for each
for (clust in unique(sig.ln.markers$cluster)){
  name <- paste("cluster", clust, "rankedGenes", sep="")
  rankedGenes <- sig.ln.markers %>%
    dplyr::filter(cluster == clust) %>%
    dplyr::select(gene, avg_log2FC) %>%
    arrange(desc(avg_log2FC))
  rankedGeneList <- rankedGenes$avg_log2FC
  names(rankedGeneList) <- rankedGenes$gene
  assign(name, rankedGeneList)
}

# import gmt file(s)
## one gmt file
#gmt <- read.gmt("HumanlnusClusters.gmt")

## multiple gmt files
gmtFiles <- list.files(pattern = "\\.gmt", full.names = TRUE) # Get list of all .gmt files in this directory
gmtTables <- lapply(gmtFiles, read.gmt) # Apply the read.gmt function to the list of .gmt files and save as a variable
gmt <- do.call(rbind, gmtTables) # Rbind files

# loop through each cluster's ranked gene list and run GSEA for gene lists in the gmt file(s) on each
for (geneList in ls(pattern = "cluster")){
  name <- paste(geneList, "GSEA", sep="_")
  input <- get(geneList)
  gse <- GSEA(input,
              exponent = 1,
              pvalueCutoff = 1,
              pAdjustMethod = "BH",
              TERM2GENE = gmt,
              verbose = TRUE,
              by = "fgsea")
  assign(name, gse)
}

# Remove any objects with no significantly enriched gene sets from the global environment
for (GSEAresult in ls(pattern = "_GSEA")){
  obj <- get(GSEAresult)
  res <- obj@result
  if (nrow(res) == 0) {
    rm(list = ls(pattern = GSEAresult))
  }
}


```

```{r GSEAplots, message=FALSE, fig.width = 20, fig.height = 16, warning=FALSE}
# Loop through each GSEA object, export results as csv, and plot results as an enrichment dot plot
plot_list <- list() # initiate empty list for individual plots

for (GSEAresult in ls(pattern = "_GSEA")){
  name <- paste("ln", GSEAresult, sep="_")
  obj <- get(GSEAresult)
  write.csv(obj, file=paste(name, "csv", sep="."))
  
  eplot <- obj %>%
    dotplot(showCategory = 10, x = "NES") +
    scale_color_viridis_c(name = "Adjusted\nP-value",
                          option = "H") +
    scale_y_discrete(labels = ~ str_wrap(gsub('_', ' ', .x), 40)) +
    geom_vline(xintercept = 0, linetype = 2) +
    labs(title = name, y = "Gene Set") +
    theme(plot.title = element_text(hjust = 0.5))
  
  plot_list[[name]] <- eplot # add plot to list
}

# combine plots into single layout
combined_plot <- wrap_plots(plot_list, ncol = 3)
print(combined_plot)
```

# Cell-Cycle Scoring
See if cluster differences are associated with differences in cell proliferation.

## Assign Cell-Cycle Scores
Each cell will be assigned a score based on its expression of G2/M and S phase markers. The CellCycleStoring() function stores S and G2/M scores in object metadata, along with the predicted classification of each cell in either G2/M, S, or G1 phase.
```{r cell-cycle}
# segregate G2/M phase and S phase markers from Seurat's built-in list of cell cycle markers
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

# assign each cell a score based on expression of G2/M and S phase markers
ln_seurat_cycle <- CellCycleScoring(ln_seurat,
                                s.features = s.genes, 
                                g2m.features = g2m.genes, 
                                set.ident = TRUE)
```

## Visualize cycling cells on UMAP
```{r CellCycleScoringUMAP-LN}
ln_seurat_cycle <- RunPCA(ln_seurat_cycle, features = c(s.genes, g2m.genes))
DimPlot(ln_seurat_cycle,
        group.by = "Phase",
        reduction = "umap",
        pt.size=0.5) + 
  ggtitle("Cell Cycle Scoring: Canine ln")
```

# Module Scoring for Conde et al. human immune cell atlas
Module scoring the average expression levels of each cluster on a single-cell level, subtracted by the aggregated expression of control feature sets. All analyzed features are binned based on averaged expression, and the control features are randomly selected from each bin.
Source: https://cells.ucsc.edu/?ds=pan-immune+global. The top 51 markers were exported from UCSC Cell Browser for each annotated cell type.

```{r moduleScoring-top51condeEtAlHumanImmuneAtlas, warning = FALSE, message = FALSE, fig.width = 16, fig.height = 12}
# make gene lists
human_ln_atlas_top51 <- read.csv("CondeEtAl_HumanImmuneCellAtlas_Global.csv")
colnames(human_ln_atlas_top51)

plot_list <- list() # initiate empty list for individual plots

for (column in colnames(human_ln_atlas_top51)){
  genes <- list(human_ln_atlas_top51[[column]])
  name <- paste(column, "Score", sep="_")
  ln_seurat <- AddModuleScore(ln_seurat, features = genes, name = name)
  newname <- paste(name, "1", sep="")
  modscoreplot <- FeaturePlot(ln_seurat, features = newname) + scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
  plot_list[[name]] <- modscoreplot # add plot to list
}

# split plot_list into groups of 6 and handle any remaining plots
plot_groups <- split(plot_list, ceiling(seq_along(plot_list) / 6))

# combined plot with dynamic number of columns/rows
combine_plots <- function(group, ncol = 3){
  n_plots <- length(group)
  nrow <- ceiling(n_plots / ncol)
  wrap_plots(group, ncol = ncol, nrow = nrow)
}
combined_plots <- lapply(plot_groups, function(group) {
  combine_plots(group, ncol = 3)
})

# display each combined plot
for (i in seq_along(combined_plots)) {
  print(combined_plots[[i]])
}
```

# Module scoring for Ammons canine circulating leukocyte atlas
Source: https://cells.ucsc.edu/?ds=canine-leukocyte-atlas+healthy 

```{r moduleScoring-top51ammonsCanineCircLeukAtlas, warning = FALSE, message = FALSE, fig.width = 16, fig.height = 12}
# make gene lists
ammonsk9_ln_atlas_top51 <- read.csv("Ammons_CanineCirculatingLeukocyteAtlas.csv")
colnames(ammonsk9_ln_atlas_top51)

plot_list <- list() # initiate empty list for individual plots

for (column in colnames(ammonsk9_ln_atlas_top51)){
  genes <- list(ammonsk9_ln_atlas_top51[[column]])
  name <- paste(column, "Score", sep="_")
  ln_seurat <- AddModuleScore(ln_seurat, features = genes, name = name)
  newname <- paste(name, "1", sep="")
  modscoreplot <- FeaturePlot(ln_seurat, features = newname) + scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
  plot_list[[name]] <- modscoreplot # add plot to list
}

# split plot_list into groups of 6 and handle any remaining plots
plot_groups <- split(plot_list, ceiling(seq_along(plot_list) / 6))

# combined plot with dynamic number of columns/rows
combine_plots <- function(group, ncol = 3){
  n_plots <- length(group)
  nrow <- ceiling(n_plots / ncol)
  wrap_plots(group, ncol = ncol, nrow = nrow)
}
combined_plots <- lapply(plot_groups, function(group) {
  combine_plots(group, ncol = 3)
})

# display each combined plot
for (i in seq_along(combined_plots)) {
  print(combined_plots[[i]])
}
```

# Assigning cell type identity to clusters
To be done once confident cluster assignment has been achieved.
```{r clustID-ln}
new.ln.cluster.ids <- c("LN0_Tcell", "LN1_BCell", "LN2_Tcell", "LN3_Tcell", "LN4_CD8_NK", "LN5_BCell", "LN6_Cycling", "LN7_Myeloid", "LN8_PlasmaCell")
names(new.ln.cluster.ids) <- levels(ln_seurat)
ln_seurat <- RenameIdents(ln_seurat, new.ln.cluster.ids)
DimPlot(ln_seurat, reduction = "umap", label = TRUE, label.size = 3, label.box = TRUE, pt.size = 0.5) + ggtitle("Canine Lymph Node, Resolution: 0.1")
saveRDS(ln_seurat, file="LN_Annotated.RData")
```

# Citations
```{r citations}
sessionInfo()
citation()
```
