---
title: "Merging T Cell Clusters"
author: "Eileen Owens"
date: "2024-12-05"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(dev = "png",
                      dpi = 300,
                      echo = TRUE,
                      cache = TRUE,
                      cache.lazy = FALSE,
                      warning = FALSE,
                      keep_md = TRUE)
```

# Introduction
The purpose of this script is to subset T-cell clusters from a canine thymus Seurat object and canine lymph node Seurat object, and integrate those T-cell clusters together into one Seurat object for downstream analysis.

# Software packages
```{r software-load, results="hide", message=FALSE}
library(Seurat)
library(knitr)
library(tidyverse)
library(clustree)
library(stringr)
library(DoubletFinder)
library(patchwork)
library(scales)
library(cowplot)
library(ggrepel)
library(colorspace)
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(SeuratDisk)
library(SingleR)
library(viridis)
library(reshape)
library(lemon)
library(ggsankey)
library(msigdbr)
library(clusterProfiler)
library(slingshot)
library(ggpubr)
library(scRNAseq)
library(scuttle)
library(ape)
library(ggtree)
```

# Data
## Working directory
```{r setwd}
setwd("C:/Users/edlarsen/Documents/240828_scRNAseq/T_Cells")
```

## Import filtered, normalized, and clustered Seurat objects
```{r load-data, fig.width = 8, fig.height = 6}
ln_seurat <- readRDS(file = "../Cluster_Annotation/LN_Annotated.RData")
thym_seurat <- readRDS(file = "../Cluster_Annotation/THYM_Annotated.RData")

# Check original cluster IDs
kable(table(Idents(ln_seurat)))
DimPlot(ln_seurat,
        reduction = "umap",
                   label = TRUE,
                   label.size = 6) + 
  plot_annotation(title = "Canine Lymph Node, Resolution: 0.1", theme = theme(plot.title = element_text(hjust = 0.5, size = 20)))

kable(table(Idents(thym_seurat)))
DimPlot(thym_seurat,
        reduction = "umap",
                   label = TRUE,
                   label.size = 6) + 
  plot_annotation(title = "Canine Thymus, Resolution: 0.2", theme = theme(plot.title = element_text(hjust = 0.5, size = 20)))
```


# Assign dataset ID and preserve original cluster labels in metadata
```{r dataset-id}
ln_seurat$dataset = "LymphNodeDataset"
thym_seurat$dataset = "ThymDataset"

ln_seurat$OriginalClusters <- Idents(ln_seurat)
thym_seurat$OriginalClusters <- Idents(thym_seurat)
```

# Subset T-cell clusters from each object
```{r subsetT}
ln_Tcells <- subset(x = ln_seurat, idents = c("LN0_Tcell", "LN2_Tcell", "LN3_Tcell", "LN4_CD8_NK"))
thym_Tcells <- subset(x = thym_seurat, idents = c("T0_DP", "T1_DP", "T2_EarlySP", "T3_DN", "T4_Cycling", "T5_CD4SP", "T6_CD8SP", "T8_Cycling"))
```

# Merge T-cell clusters into one object
```{r merge}
all_Tcells <- merge(x = ln_Tcells, y = thym_Tcells, project = "MergedTCells")
all_Tcells
kable(table(Idents(all_Tcells)))
```

# Normalize and scale data
```{r normalizeAndScale, fig.width = 8, fig.height = 6, warning=FALSE}
# normalize data
all_Tcells <- NormalizeData(all_Tcells,
                            normalization.method = "LogNormalize",
                            scale.factor = 10000)

# find variable features
all_Tcells <- FindVariableFeatures(all_Tcells)
# Make table of 10 most highly variable genes
top10_Tcells <- head(VariableFeatures(all_Tcells), 10)
plot1 <- VariableFeaturePlot(all_Tcells)
plot2 <- LabelPoints(
  plot = plot1,
  points = top10_Tcells,
  repel = TRUE
) + 
  labs(title = "Top 10 Most Variable Features:\n Merged Canine T-cells from Lymph Node and Thymus")
plot2

# scale data
all_Tcells <- ScaleData(all_Tcells)
```

# Linear dimensional reduction (PCA)
```{r pca, message=FALSE, fig.width = 8, fig.height = 6}
all_Tcells <- RunPCA(all_Tcells)
DimPlot(all_Tcells, reduction = "pca") + ggtitle("PCA: Merged Canine T-cells from Lymph Node and Thymus")
```

```{r 15pc, message=FALSE, fig.width = 8, fig.height = 10}
DimHeatmap(all_Tcells, dims = 1:15, cells = 500, balanced = TRUE)
```

## Elbow plot
To decide how many principal components to include, we rank principal components based on the percentage of variance explained by an "elbow" in the plot.
```{r ElbowPlot}
ElbowPlot(all_Tcells, ndims=40) + ggtitle("Elbow plot of principal components: \nMerged Canine T-cells from Lymph Node and Thymus")
```
The location of the elbow in this plots suggests that the majority of the true signal is captured in the first 10 PCs. Based on this, the first 10 PCs will be used to generate cell clusters.

# Perform integration
Canonical correlation analysis (CCA) is a form of PCA, except it only identifies the greatest sources of variation in the data if it is shared or conserved across the conditions/groups, using the 3000 most variable genes from each sample.
```{r integrate}
all_Tcells <- IntegrateLayers(object = all_Tcells,
                              method = CCAIntegration,
                              orig.reduction = "pca",
                              new.reduction = "integrated.cca",
                              verbose = FALSE)
all_Tcells[["RNA"]] <- JoinLayers(all_Tcells[["RNA"]])
```

```{r}
saveRDS(all_Tcells, "integrated_Tcells.RData")
```



# Cluster cells
```{r clustering, message=FALSE, results='hide', message=FALSE}
all_Tcells <- FindNeighbors(all_Tcells, dims = 1:10) # replace with PCs calculated above
all_Tcells <- FindClusters(all_Tcells, resolution = c(0.1,0.15,0.2,0.3,0.4,0.45,0.5,0.6,0.7,0.8,0.9,1.0,1.1)) # use a value above/below 1.0 if you want to obtain a larger/smaller number of clusters (communities); setting this parameter between 0.4-1.2 typically returns good results for single-cell datasets of ~3K cells; optimal resolution often increases for larger datasets.

# Check for any columns matching the "RNA_snn_res." prefix for resolutions not in the list above, which may be leftover from prior analyses, and remove these prior to running clustree (will throw an error otherwise)
resolutions <- c(0.1,0.15,0.2,0.3,0.4,0.45,0.5,0.6,0.7,0.8,0.9,1.0,1.1) # should match FindClusters code above
cols_to_keep <- paste0("RNA_snn_res.", resolutions)
rna_snn_cols <- grep("^RNA_snn_res\\.", names(all_Tcells[[]]), value = TRUE)
cols_to_remove <- setdiff(rna_snn_cols, cols_to_keep)
## Remove
all_Tcells[[cols_to_remove]] <- NULL
```

## clustree
```{r clustree, fig.width=12, fig.height=12}
clustreeT <- clustree(all_Tcells, prefix = "RNA_snn_res.")
clustreeT + ggtitle("Merged Canine T-cells from Lymph Node and Thymus") + theme(plot.title = element_text(hjust = 0.5))
```


## Non-linear Dimensional Reduction (UMAP)
```{r umap-Dims01to04, fig.width = 16, fig.height= 14}
all_Tcells <- RunUMAP(all_Tcells, dims = 1:10) # replace with the PCs determined from elbow plot above
umapT <- DimPlot(all_Tcells,
                 reduction = "umap",
                 ncol = 2,
                 group.by=glue::glue("RNA_snn_res.{c(0.1,0.15,0.2,0.3)}"),
                 label = TRUE,
                 label.size = 6)
umapT + plot_annotation(title = "Merged Canine T-cells from Lymph Node and Thymus", theme = theme(plot.title = element_text(hjust = 0.5, size = 20)))
```

```{r UMAP-LN-Dims045to06, message=FALSE, results='hide', fig.width = 16, fig.height= 14}
umapT <- DimPlot(all_Tcells,
                 reduction = "umap",
                 ncol = 2,
                 group.by=glue::glue("RNA_snn_res.{c(0.4,0.45,0.5,0.6)}"),
                 label = TRUE,
                 label.size = 6)
umapT + plot_annotation(title = "Merged Canine T-cells from Lymph Node and Thymus", theme = theme(plot.title = element_text(hjust = 0.5, size = 20)))
```

## Compare to original cluster IDs
```{r, fig.width = 10, fig.hight = 8}
DimPlot(all_Tcells, 
        reduction = "umap",
        group.by = "OriginalClusters",
        label = TRUE,
        label.size = 3,
        label.box = TRUE)
```



# Annotation
## Differentially expressed features
```{r FindAllMarkers, message=FALSE, fig.width = 8, fig.height = 6}
all_Tcells <- FindClusters(all_Tcells, resolution = 0.3)
all_Tcells <- RunUMAP(all_Tcells, dims = 1:10)
Tcell.markers <- FindAllMarkers(all_Tcells, only.pos=TRUE)

# Export
write.csv(Tcell.markers, file="mergedLnAndThymTCells_DEfeaturesByUMAPcluster_0.3.csv")
```

## Top cluster markers
```{r topClusterMarkers, fig.width = 18, fig.height = 8}
key.genes <- Tcell.markers[!grepl("ENSCAFG", row.names(Tcell.markers)),]
key.genes.sortedByPval <- key.genes[order(key.genes$p_val),]
features <- key.genes.sortedByPval %>% group_by(cluster) %>% do(head(., n=5))
features <- as.data.frame(features[!duplicated(features$gene),]) # prioritize unique clusters

DotPlot(all_Tcells,
        assay = "RNA",
        features = rev(features$gene),
        scale = TRUE) + 
  geom_point(aes(size=pct.exp), shape = 21, colour="black", stroke=0.5) +
  labs(y = "Cluster") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```

## Feature plots

### Early thymic progenitors
In humans and mice, the earliest lymphoid progenitor cells express **c-kit**, **Flt3**, **CD27**, and variably **IL7Ra**. T-cell lineage commitment is facilitated by **Notch** signaling in the thymus, which triggers expression of transcription factors T cell factor 1 (TCF-1, encoded by **TCF7**), **GATA3**, and **BCL11b** that promote expression of genes required for T-cell survival and differentiation.

```{r FeaturePlot-Early, fig.width=20, fig.height=14}
FeaturePlot(all_Tcells,
            reduction = "umap",
            features = c("CD34", "KIT", "FLT3", "IL7R", "NOTCH1", "TCF7", "GATA3", "BCL11B")) +
  plot_annotation(title = "Merged Canine T-cells from Lymph Node and Thymus:\nExpression of Early Thymocyte Progenitor (ETP) Markers", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```

### Double negative thymocytes
DN thymocytes should lack expression of CD4 and CD8. Four stages comprise the double negative (DN) thymocytes, distinguished by their variable expression of **CD44** and **CD25 (IL2RA)**. **Notch** signaling drives the progression from **CD25-CD44+ DN1** thymocytes to **CD25+CD44+ DN2** thymocytes. After being downregulated in these late DN stages, CD44 will again be expressed in SP thymocytes and is also a common marker of activation of mature T cells.

Between the DN2 and DN3 stage, thymocytes rearrange their TCR to become committed to either the α/β or γ/δ T-cell lineage. Rearrangement of the TCR is facilitated by enzymes **RAG1** and **RAG2**. The TCRβ chain is rearranged first, and prior to TCRα chain rearrangement, it associates with the invariant **pre-Tα (PTCRA)** chain and CD3 molecules to form the pre-TCR. DN3 cells that pass β-selection will lose expression of CD25 and CD44 to become DN4 cells.

In cynomolgus monkeys, **CD1** is expressed in DN and DP thymocytes, then downregulated as cells progressed to the SP stages.

In mice, **LY6E** (also known as Thymic Shared Antigen-1 or Stem Cell Antigen-2) is a marker distinguishing early DN stages from more mature thymocyte subsets. In mature T cells, LY6E is a marker for T-cell activation.

```{r FeaturePlot-DN, fig.width=20, fig.height=12}
FeaturePlot(all_Tcells,
            reduction = "umap",
            features = c("CD4", "CD8A", "CD44", "IL2RA", "NOTCH1", "RAG1", "RAG2", "PTCRA", "DNTT", "CD1C", "LY6E")) +
  plot_annotation(title = "Merged Canine T-cells from Lymph Node and Thymus:\nExpression of Double Negative (DN) Thymocyte Markers", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```

DN4 thymocytes undergo a proliferative burst, which may be identified by cell cycle scoring to identify proliferating cells:

```{r cell-cycle}
# segregate G2/M phase and S phase markers from Seurat's built-in list of cell cycle markers
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

# assign each cell a score based on expression of G2/M and S phase markers
all_Tcells_cycle <- CellCycleScoring(all_Tcells,
                                s.features = s.genes, 
                                g2m.features = g2m.genes, 
                                set.ident = TRUE)

all_Tcells_cycle <- RunPCA(all_Tcells_cycle, features = c(s.genes, g2m.genes))
DimPlot(all_Tcells_cycle,
        group.by = "Phase",
        reduction = "umap",
        pt.size=0.5) + 
  ggtitle("Cell Cycle Scoring:\nMerged Canine T-cells from Lymph Node and Thymus")
```

```{r ki67}
FeaturePlot(all_Tcells,
            reduction = "umap",
            features = c("MKI67")) +
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```

### Double positive thymocytes

DP thymocytes should express CD4 and CD8. **CCR9** is expressed by most DP thymocytes, but not DN thymocytes, and it is subsequently downregulated with the transition from the DP stage to SP stages of thymocyte development. In cynomolgus monkeys, **CD1** is expressed in DN and DP thymocytes, then downregulated as cells progressed to the SP stages.

```{r FeaturePlot-DP, fig.width=14, fig.height=10}
FeaturePlot(all_Tcells,
            reduction = "umap",
            features = c("CD4", "CD8A", "CCR9", "CD1C")) +
  plot_annotation(title = "Merged Canine T-cells from Lymph Node and Thymus:\nExpression of Double Positive Thymocyte Markers", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```

### Single positive thymocytes

SP thymocytes should express either CD4 (mediated by transcription factors **GATA3** and THPOK, encoded by **ZBTB7B**) or CD8 (mediated by transcription factor **RUNX3**). The **IL-7/STAT5** axis acts upstream of Runx3 to enhance its expression and promote CD8+ differentiation. During terminal maturation, CD4+ and CD8+ T cells start expressing sphingosine 1 phosphate receptor (**S1PR1**) which promotes egress from the thymus and entry into circulation.

```{r FeaturePlot-SP, fig.width=20, fig.height=16}
FeaturePlot(all_Tcells,
            reduction = "umap",
            ncol = 3,
            features = c("CD4", "CD8A", "GATA3", "ZBTB7B", "RUNX3", "STAT5A", "S1PR1")) +
  plot_annotation(title = "Merged Canine T-cells from Lymph Node and Thymus:\nExpression of Double Positive Thymocyte Markers", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```

### Naive T cells

The homing molecule L-selectin (encoded by **SELL**) is primarily expresed by naive T cells and is downregulated following engagement of the TCR by antigen. Another homing molecule expressed by naive T cells is **CCR7**. CD132 (**IL2RG**) is expressed by both mature and activated mature T cells, while CD122 (**IL2RB**) tends to be expressed more highly in activated T cells than naive T cells. 

```{r FeaturePlot-Naivemarkers, fig.width = 14, fig.height = 10}
FeaturePlot(all_Tcells, 
            reduction = "umap",
            ncol = 2,
            features = c("SELL", "CCR7", "IL2RG", "IL2RB")) +
  plot_annotation(title = "Merged Canine T-cells from Lymph Node and Thymus:\nExpression of Markers of Naive T Cells", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```

These markers for naive CD4+ and CD8+ T cells were derived from a single-cell atlas of circulating canine leukocytes (https://github.com/dyammons/Canine_Leukocyte_scRNA).

```{r FeaturePlot-NaiveCD8markers, fig.height = 16, fig.width = 14}
FeaturePlot(all_Tcells,
            reduction = "umap",
            features = c("CD8A", "ITGA1", "PTGDR", "IL2RB", "ADGRG1", "NBEA")) + 
  plot_annotation(title = "Merged Canine T-cells from Lymph Node and Thymus:\nExpression of Markers for Naive CD8 T-cells from Canine Circulating Leukocyte Atlas", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```

```{r FeaturePlot-NaiveCD4markers, fig.height = 14, fig.width = 16}
FeaturePlot(all_Tcells,
            reduction = "umap",
            features = c("CD4", "LEF1", "CSTA", "RGS10", "ZNF536", "CCR7", "COL6A5", "LTB", "TNFSF8")) + 
  plot_annotation(title = "Merged Canine T-cells from Lymph Node and Thymus:\nExpression of Markers for Naive CD4 T-cells from Canine Circulating Leukocyte Atlas", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```

### Mature, activated T cells
In mature T cells, **LY6E** is a marker for T-cell activation. T cells rapidly upregulate adhesion molecule **CD44** after antigen encounter/activation. The culmination of T-cell activation is their clonal expansion as a result of their secretion of **IL-2**. T cells increasingly upregulate inhibitory molecules **CTLA-4** and **LGALS3** after stimulation. **CD7** is a marker for mature T cells but is also expressed by NK cells. 

```{r FeaturePlot-Maturemarkers, fig.width = 20, fig.height = 12}
FeaturePlot(all_Tcells, 
            reduction = "umap",
            ncol = 3,
            features = c("CD3E", "CD7", "IL2", "IL2RA", "CTLA4", "LGALS3")) +
  plot_annotation(title = "Merged Canine T-cells from Lymph Node and Thymus:\nExpression of Markers of T-cell Maturity", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```
### Mature T-cell subtypes

```{r FeaturePlot-Th1markers, fig.width = 16, fig.height = 12}
FeaturePlot(all_Tcells, 
            reduction = "umap",
            ncol = 3,
            features = c("CD4", "TBX21", "STAT4", "STAT1", "IFNG", "IL12RB1", "IL12RB2", "IL2", "TNFA", "TNFB", "CXCR3")) +
  plot_annotation(title = "Merged Canine T-cells from Lymph Node and Thymus:\nExpression of Canonical Th1 Markers", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```

```{r FeaturePlot-Th2markers, fig.width = 16, fig.height = 8}
FeaturePlot(all_Tcells, 
            reduction = "umap",
            ncol = 3,
            features = c("CD4", "GATA3", "STAT6", "IL4", "IL4R", "CCR4")) +
  plot_annotation(title = "Merged Canine T-cells from Lymph Node and Thymus:\nExpression of Canonical Th2 Markers", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```

```{r FeaturePlot-Tregmarkers, fig.width = 10, fig.height = 8}
FeaturePlot(all_Tcells, 
            reduction = "umap",
            ncol = 2,
            features = c("CD4", "IL2RA", "FOXP3", "IL10")) +
  plot_annotation(title = "Merged Canine T-cells from Lymph Node and Thymus:\nExpression of Canonical Treg Markers", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```

```{r FeaturePlot-Th17markers, fig.width = 16, fig.height = 12}
FeaturePlot(all_Tcells, 
            reduction = "umap",
            ncol = 3,
            features = c("CD4", "RORC", "IL22", "IL23R", "IL23A", "AHR", "TGFBR1", "TGFBR2", "STAT3")) +
  plot_annotation(title = "Merged Canine T-cells from Lymph Node and Thymus:\nExpression of Canonical Th17 Markers", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```


```{r FeaturePlot-Th9markers, fig.width = 10, fig.height = 8}
FeaturePlot(all_Tcells, 
            reduction = "umap",
            ncol = 2,
            features = c("CD4", "IRF4", "SPI1", "STAT6")) +
  plot_annotation(title = "Merged Canine T-cells from Lymph Node and Thymus:\nExpression of Canonical Th9 Markers", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```

```{r FeaturePlot-Th22markers, fig.width = 10, fig.height = 8}
FeaturePlot(all_Tcells, 
            reduction = "umap",
            ncol = 2,
            features = c("CD4", "AHR", "RORC", "IL22")) +
  plot_annotation(title = "Merged Canine T-cells from Lymph Node and Thymus:\nExpression of Canonical Th22 Markers", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```

```{r FeaturePlot-CD8markers, fig.width = 14, fig.height = 8}
FeaturePlot(all_Tcells, 
            reduction = "umap",
            ncol = 3,
            features = c("CD8A", "IFNG", "TNFA", "FAS", "FASLG", "GZMB", "PRF1")) +
  plot_annotation(title = "Merged Canine T-cells from Lymph Node and Thymus:\nExpression of Canonical CD8+ T-cell Markers", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```

```{r FeaturePlot-NKmarkers, fig.width = 16, fig.height = 12}
FeaturePlot(all_Tcells, 
            reduction = "umap",
            ncol = 3,
            features = c("NCAM1", "KLRB1", "FCER1G", "CD7", "CD27", "GZMA", "GZMB")) +
  plot_annotation(title = "Merged Canine T-cells from Lymph Node and Thymus:\nExpression of Canonical NK Cell Markers", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```

# Module scoring
## Park et al. human thymus feature expression
Calculates the average expression levels of each cluster on a single-cell level, subtracted by the aggregated expression of control feature sets. All analyzed features are binned based on averaged expression, and the control features are randomly selected from each bin.

### Top 20 marker genes for each cell type in human thymus atlas
Source: Park et al., Supplementary Table 4 (https://www.science.org/doi/10.1126/science.aay3224#supplementary-materials)
```{r moduleScoring-top20parkEtAlHumanThymusAtlas, warning = FALSE, fig.width = 16, fig.height = 8, message = FALSE}
# make gene lists
human_thymus_atlas_top20 <- read.csv("../Cluster_Annotation/Parketal_SupTable4_Top20MarkersForCellTypes.csv")
colnames(human_thymus_atlas_top20)

plot_list <- list() # initiate empty list for individual plots

for (column in colnames(human_thymus_atlas_top20)){
  genes <- list(human_thymus_atlas_top20[[column]])
  name <- paste(column, "Score", sep="_")
  all_Tcells <- AddModuleScore(all_Tcells, features = genes, name = name)
  newname <- paste(name, "1", sep="")
  plotname <- paste(name, "modscoreplot", sep="_")
  modscoreplot <- FeaturePlot(all_Tcells, features = newname) + scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
  plot_list[[name]] <- modscoreplot # add plot to list
}

# split plot_list into groups of 4 and handle any remaining plots
plot_groups <- split(plot_list, ceiling(seq_along(plot_list) / 6))

# combined plot with dynamic number of columns/rows
combine_plots <- function(group, ncol = 3){
  n_plots <- length(group)
  nrow <- ceiling(n_plots / ncol)
  wrap_plots(group, ncol = ncol, nrow = nrow)
}
combined_plots <- lapply(plot_groups, function(group) {
  combine_plots(group, ncol = 3)
})

# display each combined plot
for (i in seq_along(combined_plots)) {
  print(combined_plots[[i]])
}
```

### Top 51 marker genes for each cell type in human thymus atlas
Source: Park et al. data accessed through UCSC Cell Browser
```{r moduleScoring-top51parkEtAlHumanThymusAtlas, warning = FALSE, message = FALSE, fig.width = 18, fig.height = 12}
# make gene lists
human_thymus_atlas_top51 <- read.csv("../Cluster_Annotation/Parketal_Top51DEGenesForThymusClusters_fromUCSCCellBrowser.csv")
colnames(human_thymus_atlas_top51)

plot_list <- list() # initiate empty list for individual plots

for (column in colnames(human_thymus_atlas_top51)){
  genes <- list(human_thymus_atlas_top51[[column]])
  name <- paste(column, "Score", sep="_")
  all_Tcells <- AddModuleScore(all_Tcells, features = genes, name = name)
  newname <- paste(name, "1", sep="")
  modscoreplot <- FeaturePlot(all_Tcells, features = newname) + scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
  plot_list[[name]] <- modscoreplot # add plot to list
}

# split plot_list into groups of 6 and handle any remaining plots
plot_groups <- split(plot_list, ceiling(seq_along(plot_list) / 6))

# combined plot with dynamic number of columns/rows
combine_plots <- function(group, ncol = 3){
  n_plots <- length(group)
  nrow <- ceiling(n_plots / ncol)
  wrap_plots(group, ncol = ncol, nrow = nrow)
}
combined_plots <- lapply(plot_groups, function(group) {
  combine_plots(group, ncol = 3)
})

# display each combined plot
for (i in seq_along(combined_plots)) {
  print(combined_plots[[i]])
}
```

## Conde et al. human immune cell atlas
Module scoring the average expression levels of each cluster on a single-cell level, subtracted by the aggregated expression of control feature sets. All analyzed features are binned based on averaged expression, and the control features are randomly selected from each bin.
Source: https://cells.ucsc.edu/?ds=pan-immune+global. The top 51 markers were exported from UCSC Cell Browser for each annotated cell type.

```{r moduleScoring-top51condeEtAlHumanImmuneAtlas, warning = FALSE, message = FALSE, fig.width = 16, fig.height = 12}
# make gene lists
human_ln_atlas_top51 <- read.csv("../Cluster_Annotation/CondeEtAl_HumanImmuneCellAtlas_Global.csv")
colnames(human_ln_atlas_top51)

plot_list <- list() # initiate empty list for individual plots

for (column in colnames(human_ln_atlas_top51)){
  genes <- list(human_ln_atlas_top51[[column]])
  name <- paste(column, "Score", sep="_")
  all_Tcells <- AddModuleScore(all_Tcells, features = genes, name = name)
  newname <- paste(name, "1", sep="")
  modscoreplot <- FeaturePlot(all_Tcells, features = newname) + scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
  plot_list[[name]] <- modscoreplot # add plot to list
}

# split plot_list into groups of 6 and handle any remaining plots
plot_groups <- split(plot_list, ceiling(seq_along(plot_list) / 6))

# combined plot with dynamic number of columns/rows
combine_plots <- function(group, ncol = 3){
  n_plots <- length(group)
  nrow <- ceiling(n_plots / ncol)
  wrap_plots(group, ncol = ncol, nrow = nrow)
}
combined_plots <- lapply(plot_groups, function(group) {
  combine_plots(group, ncol = 3)
})

# display each combined plot
for (i in seq_along(combined_plots)) {
  print(combined_plots[[i]])
}
```

## Ammons canine circulating leukocyte atlas
Source: https://cells.ucsc.edu/?ds=canine-leukocyte-atlas+healthy 

```{r moduleScoring-top51ammonsCanineCircLeukAtlas, warning = FALSE, message = FALSE, fig.width = 16, fig.height = 12}
# make gene lists
ammonsk9_ln_atlas_top51 <- read.csv("../Cluster_Annotation/Ammons_CanineCirculatingLeukocyteAtlas.csv")
colnames(ammonsk9_ln_atlas_top51)

plot_list <- list() # initiate empty list for individual plots

for (column in colnames(ammonsk9_ln_atlas_top51)){
  genes <- list(ammonsk9_ln_atlas_top51[[column]])
  name <- paste(column, "Score", sep="_")
  all_Tcells <- AddModuleScore(all_Tcells, features = genes, name = name)
  newname <- paste(name, "1", sep="")
  modscoreplot <- FeaturePlot(all_Tcells, features = newname) + scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
  plot_list[[name]] <- modscoreplot # add plot to list
}

# split plot_list into groups of 6 and handle any remaining plots
plot_groups <- split(plot_list, ceiling(seq_along(plot_list) / 6))

# combined plot with dynamic number of columns/rows
combine_plots <- function(group, ncol = 3){
  n_plots <- length(group)
  nrow <- ceiling(n_plots / ncol)
  wrap_plots(group, ncol = ncol, nrow = nrow)
}
combined_plots <- lapply(plot_groups, function(group) {
  combine_plots(group, ncol = 3)
})

# display each combined plot
for (i in seq_along(combined_plots)) {
  print(combined_plots[[i]])
}
```

# Assigning cell type identity to clusters
To be done once confident cluster assignment has been achieved.
```{r clustID, fig.width = 10, fig.height = 8}
new.cluster.ids <- c("Naive_T", "DP_Thymocytes_2", "Activated_T", "DP_Thymocytes_3", "SP_Thymocytes_1", "SP_Thymocytes_2", "DP_Thymocytes_1", "DN_Thymocytes", "CD8_NK_1", "NK_cells", "Unknown", "CD8_NK_2")
names(new.cluster.ids) <- levels(all_Tcells)
all_Tcells <- RenameIdents(all_Tcells, new.cluster.ids)

DimPlot(all_Tcells, reduction = "umap", label = TRUE, label.size = 3, label.box = TRUE, pt.size = 0.5) + ggtitle("Canine Merged Lymph Node and Thymus T Cells, Resolution: 0.3")

# save as new column in metadata in addition to reassigning idents
all_Tcells$IntegratedClusters <- Idents(all_Tcells)
all_Tcells$IntegratedClusters <- Idents(all_Tcells)

# export
#saveRDS(all_Tcells, file="IntegThymAndLN_Annotated.RData")
```

# Citations
```{r citations}
sessionInfo()
citation()
```

