---
title: "Seurat Data Normalization And Clutering: Integrated scRNA-seq Data"
author: "Eileen Owens"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(dev = "png",
                      dpi = 300,
                      echo = TRUE,
                      cache = TRUE,
                      warning = FALSE,
                      keep_md = TRUE)
```

# Introduction
The purpose of this script is to perform single-cell RNA-seq data normalization and clustering with Seurat. This follows the standard pre-processing workflow as described in the the Seurat documentation https://satijalab.org/seurat/articles/pbmc3k_tutorial#normalizing-the-data.

# Loading data and packages
## Load necessary software packages
```{r software-load, results="hide", message=FALSE}
library(Seurat)
library(tidyverse)
library(clustree)
library(stringr)
library(DoubletFinder)
library(patchwork)
library(scales)
library(cowplot)
library(ggrepel)
library(colorspace)
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(SeuratDisk)
library(SingleR)
library(viridis)
library(reshape)
library(lemon)
library(ggsankey)
library(msigdbr)
library(clusterProfiler)
library(slingshot)
library(ggpubr)
library(scRNAseq)
library(scuttle)
library(ape)
library(ggtree)
library(presto)
library(patchwork)
```

## Set working directory
```{r setwd}
setwd("C:/Users/edlarsen/Documents/240828_scRNAseq")
```

## Load filtered Seurat objects
For this analysis, datasets that had been merged and integrated by location (lymph node or thymus) were used.
```{r}
integ_ln <- readRDS(file = "seurat_integrated_LN_filtered.RData")
integ_thym <- readRDS(file = "seurat_integrated_THYM_filtered.RData")
```

# Data normalization
The global-scaling normalization method "LogNormalize" normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000), and log-transforms the result.

```{r NormalizeData, message=FALSE, results='hide'}
# lymph node
normalized_ln <- NormalizeData(integ_ln,
                               normalization.method = "LogNormalize",
                               scale.factor = 10000)

# thymus
normalized_thym <- NormalizeData(integ_thym,
                                 normalization.method = "LogNormalize",
                                 scale.factor = 10000)
```

# Identification of highly variable features (feature selection)
Calculate a subset of features that exhibit high cell-to-cell variation in the dataset (i.e., highly expressed in some cells and lowly expressed in others). Focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets. Seurat's procedure involves directly modeling the mean-variance relationship inherent in single-cell data and returning 2,000 features per dataset by default that can be used in downstream analysis like PCA. Reference for Seurat's procedure: https://www.sciencedirect.com/science/article/pii/S0092867419305598?via%3Dihub. 

```{r FindVariableFeatures, message=FALSE, fig.fullwidth=TRUE}
normalized_ln <- FindVariableFeatures(normalized_ln) # lymph node
normalized_thym <- FindVariableFeatures(normalized_thym) # thymus

# Identify the 10 most highly variable genes
top10_ln <- head(VariableFeatures(normalized_ln), 10)
top10_thym <- head(VariableFeatures(normalized_thym), 10)

# plot variable features
plot1 <- VariableFeaturePlot(normalized_ln)

plot2 <- LabelPoints(
  plot = plot1,
  points = top10_ln,
  repel = TRUE
) + 
  labs(title = "Top 10 Most Variable Features: Canine Lymph Node")

plot2


plot3 <- VariableFeaturePlot(normalized_thym)

plot4 <- LabelPoints(
  plot = plot3,
  points = top10_thym,
  repel = TRUE
) +
  labs(title = "Top 10 Most Variable Features: Canine Thymus")

plot4 
```

# Scaling the data
Apply a linear transformation that is a standard pre-processing step prior to dimensional reduction techniques like PCA. This transformation works by shifting the expression of each gene, so that the mean expression across cells is 0. Then it scales the expression of each gene, so that the variance across cells is 1 (this step gives equal weight in downstream analyses, so that highly expressed genes do not dominate).

```{r ScaleData, message=FALSE, results='hide'}
# By default, only variable features are scaled. To evaluate all genes and not just variable features: ScaleData(data, features = all.genes)

normalized_ln <- ScaleData(normalized_ln) # lymph node
normalized_thym <- ScaleData(normalized_thym) # thymus
```

# Perform linear dimensional reduction
Each PC of a PCA score essentially represents a 'metafeature' that combines information across a correlated feature set. The top principal components therefore represent a robust compression of the dataset. 

```{r RunPCA, message=FALSE, fig.height=6}
# By default, the previously determined variable features are used as input, but can be defined using the 'features' argument; if you want to do a custom set of features, make sure they are passed to ScaleData() first.

normalized_ln <- RunPCA(normalized_ln) # lymph node
normalized_thym <- RunPCA(normalized_thym) # thymus

VizDimLoadings(normalized_ln, dims = 1:2, reduction = "pca") + ggtitle("PCA loadings: Canine Lymph Node") + theme(plot.title = element_text(hjust = -10, vjust = 3))
VizDimLoadings(normalized_thym, dims = 1:2, reduction = "pca") + ggtitle("PCA loadings: Canine Thymus") + theme(plot.title = element_text(hjust = -10, vjust = 3))
```

```{r DimPlot}
DimPlot(normalized_ln, reduction = "pca") + ggtitle("PCA: Canine Lymph Node")
DimPlot(normalized_thym, reduction = "pca") + ggtitle("PCA: Canine Thymus")
```

DimHeatmap() allows for easy exploration of the primary sources of heterogeneity in a dataset, and can be useful when trying to decide which PCs to include for further downstream analysis. Both cells and features are ordered according to their PCA scores. Setting 'cells' to a number plots the 'extreme' cells on both ends of the spectrum. Though a supervised analysis, this is a valuable tool for exploring correlated feature sets.

```{r DimHeatmap}
DimHeatmap(normalized_ln, 
           dims = 1, 
           cells = 500, 
           balanced = TRUE,
           fast = FALSE) + # fast = FALSE allows customized legends with ggplot
  ggtitle("PC_1: Canine Lymph Node") +
  theme(plot.title = element_text(hjust = 0.5))

DimHeatmap(normalized_thym, 
           dims = 1, 
           cells = 500, 
           balanced = TRUE,
           fast = FALSE) +
  ggtitle("PC_1: Canine Thymus") +
  theme(plot.title = element_text(hjust = 0.5))
```

## 15 PC dimensions: Lymph node
```{r pc15LN, fig.width=10, fig.height=12}
DimHeatmap(normalized_ln,
           dims = 1:15,
           cells = 500,
           balanced = TRUE)
```


## 15 PC dimensions: Thymus
```{r pc15THYM, fig.width=10, fig.height=12}
DimHeatmap(normalized_thym,
           dims = 1:15,
           cells = 500,
           balanced = TRUE)
```

# Determine dataset dimensionality
To decide how many principal components to include, we rank principal components based on the percentage of variance explained by an "elbow" in the plot.

```{r ElbowPlot}
ElbowPlot(normalized_ln, ndims=40) + ggtitle("Elbow plot of principal components: Canine Lymph node")
ElbowPlot(normalized_thym, ndims=40) + ggtitle("Elbow plot of principal components: Canine Thymus")
```
The location of the elbow in these plots suggests that the majority of the true signal is captured in the first 10 PCs. Based on this, the first 10 PCs will be used to generate cell clusters for the thymus and lymph node in Seurat.


# Cluster cells
Seurat first embeds cells in a K-nearest neighbor graph based on the euclidean distance in PCA space, and refines the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). This draws edges between cells with similar feature expression patterns. To cluster the cells, modularity optimization is then applied to iteratively group cells together.

The resolution parameter determines how many clusters are created. Values above 1.0 generate a larger number of clusters, and values below 1.0 generate a smaller number of clusters. 0.4-1.2 typically returns good result for single-cell datasets of ~3k cells, but optimal resolution often increases for larger datasets. This script will initially set this parameter broadly and then evaluate the stability of the clusters drawn at each resolution using clustree.

```{r clustering, message=FALSE, results='hide'}
## lymph node
normalized_ln <- FindNeighbors(normalized_ln, dims = 1:10) # replace with the PCs calculated above
normalized_ln <- FindClusters(normalized_ln, resolution = c(0.4,0.5,0.6,0.7,0.8,0.9,1.0,1.1,1.2)) # use a value above/below 1.0 if you want to obtain a larger/smaller number of clusters (communities); setting this parameter between 0.4-1.2 typically returns good results for single-cell datasets of ~3K cells; optimal resolution often increases for larger datasets.
head(normalized_ln[[]]) # clusters at the various resolutions are stored in the RNA_snn.res columns
normalized_ln[["RNA_snn_res.2"]] <- NULL # remove any additional resolutions not in the list above

## thymus
normalized_thym <- FindNeighbors(normalized_thym, dims = 1:10) # replace with the PCs calculated above
normalized_thym <- FindClusters(normalized_thym, resolution = c(0.4,0.5,0.6,0.7,0.8,0.9,1.0,1.1,1.2))
head(normalized_thym[[]])
normalized_thym[["RNA_snn_res.2"]] <- NULL # remove any additional resolutions not in the list above
```

## clustree
We want to determine the optimal resolution and cluster number to achieve an appropriate balance between over-clustering and under-clustering in order to keep cell population identities as accurate and biologically informative as possible. Seurat has parameters for tuning the resolution and number of clusters, but this does not determine whether these generated clusters are meaningful.

Drawing a clustering tree can help compare the stability of clusterings at a range of resolutions. The node size is related to the number of cells in each cluster, and the node color indicates the clustering resolution. Arrows are colored according to the number of samples they represent, and the arrow transparency shows the incoming node proportion (the number of samples in the edge divided by the number of samples in the node it points to). As the tree becomes messier and nodes have multiple incoming edges, it indicates overclustering of the data. (Note that clustree does not do any clustering of its own, but rather checks the clustering done by Seurat.)

```{r clustree, fig.width=12, fig.height=12}
clustree1 <- clustree(normalized_ln, prefix = "RNA_snn_res.")
clustree1 + ggtitle("Canine Lymph Node Clusters") + theme(plot.title = element_text(hjust = 0.5))

clustree2 <- clustree(normalized_thym, prefix = "RNA_snn_res.")
clustree2 + ggtitle("Canine Thymus Clusters") + theme(plot.title = element_text(hjust = 0.5))
```

Clusters for the canine lymph node appear most stable at the 0.5-0.7 resolution. Clusters for the canine thymus appear most stable at the 0.4-0.5 resolution.


# Run non-linear dimensional reduction (UMAP)
These algorithms learn the underlying structure in the dataset in order to place similar cells together in low-dimensional space. Cells that are grouped together within graph-based clusters determined above should co-localize on these dimension reduction plots.
All visualization techniques have limitations and cannot fully represent the complexity of the underlying data.


## Lymph Node

```{r UMAP-LN, message=FALSE, results='hide', fig.width = 16}
normalized_ln <- RunUMAP(normalized_ln, dims = 1:10) # replace with the PCs calculated above
umap_ln <- DimPlot(normalized_ln,
                   reduction = "umap",
                   ncol = 2,
                   group.by=glue::glue("RNA_snn_res.{c(0.4,0.5,0.6,0.7)}"), # plot resolutions based on clustree results
                   label = TRUE,
                   label.size = 6)
umap_ln + plot_annotation(title = "Canine Lymph Node", theme = theme(plot.title = element_text(hjust = 0.5, size = 20)))
```

## Thymus
```{r UMAP-THYM, message=FALSE, results='hide', fig.width = 12}
normalized_thym <- RunUMAP(normalized_thym, dims = 1:10) # replace with the PCs calculated above
umap_thym <- DimPlot(normalized_thym,
                     reduction = "umap",
                     ncol = 3,
                     group.by=glue::glue("RNA_snn_res.{c(0.4,0.5,0.6)}"), # plot resolutions based on clustree results
                     label = TRUE,
                     label.size = 6)
umap_thym + plot_annotation(title = "Canine Thymus", theme = theme(plot.title = element_text(hjust = 0.5, size = 20)))
```

# Find differentially expressed features
By default, identifies all positive and negative markers of a single cluster compared to all other cells.

## Lymph node
```{r FindAllMarkers-LN, message=FALSE}

# 0.4 resolution
normalized_ln_0.4 <- FindClusters(normalized_ln, resolution = 0.4)
normalized_ln_0.4 <- RunUMAP(normalized_ln_0.4, dims = 1:10)
ln.markers_0.4 <- FindAllMarkers(normalized_ln_0.4, only.pos=TRUE)
ln.markers_0.4 %>%
  group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 1)

# 0.5 resolution
normalized_ln_0.5 <- FindClusters(normalized_ln, resolution = 0.5)
normalized_ln_0.5 <- RunUMAP(normalized_ln_0.5, dims = 1:10)
ln.markers_0.5 <- FindAllMarkers(normalized_ln_0.5, only.pos=TRUE)
ln.markers_0.5 %>%
  group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 1)

# 0.6 resolution
normalized_ln_0.6 <- FindClusters(normalized_ln, resolution = 0.5)
normalized_ln_0.6 <- RunUMAP(normalized_ln_0.6, dims = 1:10)
ln.markers_0.6 <- FindAllMarkers(normalized_ln_0.6, only.pos=TRUE)
ln.markers_0.6 %>%
  group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 1)

# 0.7 resolution
normalized_ln_0.7 <- FindClusters(normalized_ln, resolution = 0.5)
normalized_ln_0.7 <- RunUMAP(normalized_ln_0.7, dims = 1:10)
ln.markers_0.7 <- FindAllMarkers(normalized_ln_0.7, only.pos=TRUE)
ln.markers_0.7 %>%
  group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 1)

# export
write.csv(ln.markers_0.4, file="241017_K9LN_DEfeaturesByUMAPcluster_10PCs_0.4res.csv")
write.csv(ln.markers_0.5, file="241017_K9LN_DEfeaturesByUMAPcluster_10PCs_0.5res.csv")
write.csv(ln.markers_0.6, file="241017_K9LN_DEfeaturesByUMAPcluster_10PCs_0.6res.csv")
write.csv(ln.markers_0.7, file="241017_K9LN_DEfeaturesByUMAPcluster_10PCs_0.7res.csv")
```


```{r FeaturePlot-LN, fig.width=20}
FeaturePlot(normalized_ln, 
            reduction = "umap",
            features = c("CCR9", "CD4", "CD8", "DNTT", "CD34", "RAG1", "RAG2", "IL7R", "CCR7", "CD14", "LYZ", "MS4A1", "CD8A", "FCGR3A", "MS4A7", "GNLY", "NKG7", "FCER1A", "CST3", "PPBP"))
```


## Thymus
```{r FindAllMarkers-THYM, message=FALSE}

# 0.4 resolution
normalized_thym_0.4 <- FindClusters(normalized_thym, resolution = 0.4)
normalized_thym_0.4 <- RunUMAP(normalized_thym_0.4, dims = 1:10)
thym.markers_0.4 <- FindAllMarkers(normalized_thym_0.4, only.pos=TRUE)
thym.markers_0.4 %>%
  group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 1)

# 0.5 resolution
normalized_thym_0.5 <- FindClusters(normalized_thym, resolution = 0.5)
normalized_thym_0.5 <- RunUMAP(normalized_thym_0.5, dims = 1:10)
thym.markers_0.5 <- FindAllMarkers(normalized_thym_0.5, only.pos=TRUE)
thym.markers_0.5 %>%
  group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 1)

# 0.6 resolution
normalized_thym_0.6 <- FindClusters(normalized_thym, resolution = 0.6)
normalized_thym_0.6 <- RunUMAP(normalized_thym_0.6, dims = 1:10)
thym.markers_0.6 <- FindAllMarkers(normalized_thym_0.6, only.pos=TRUE)
thym.markers_0.6 %>%
  group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 1)

write.csv(thym.markers_0.4, file="241008_K9THYMUS_DEfeaturesByUMAPcluster_10PCs_0.4res.csv")
write.csv(thym.markers_0.5, file="241008_K9THYMUS_DEfeaturesByUMAPcluster_10PCs_0.5res.csv")
write.csv(thym.markers_0.6, file="241008_K9THYMUS_DEfeaturesByUMAPcluster_10PCs_0.6res.csv")
```

```{r FeaturePlot-THYM, fig.width=12, fig.height=14}
FeaturePlot(normalized_thym,
            reduction = "umap",
            features = c("CCR9", "CD4", "CD8", "DNTT", "CD34", "RAG1", "RAG2", "IL7R", "CCR7", "CD14", "LYZ", "MS4A1", "CD8A", "FCGR3A", "MS4A7", "GNLY", "NKG7", "FCER1A", "CST3", "PPBP"))
```

# Export Seurat objects
```{r}
saveRDS(normalized_ln, file="seurat_NormalizedAndClustered_LN.RData")
saveRDS(normalized_thym, file="seurat_NormalizedAndClustered_THYM.RData")
```


# Citations
```{r citations}
sessionInfo()
citation()
```


