---
title: "Lymph Node Single-Cell Cluster Annotation"
author: "Eileen Owens"
date: "2024-11-04"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(dev = "png",
                      dpi = 300,
                      echo = TRUE,
                      cache = TRUE,
                      warning = FALSE,
                      keep_md = TRUE)
```

# Introduction
The purpose of this script is to annotate single-cell RNA-seq clusters following filtering, normalization, and clustering of the data with Seurat. This script has been adapted from the Seurat (https://satijalab.org/seurat/) and SingleR (https://bioconductor.org/packages/devel/bioc/vignettes/SingleR/inst/doc/SingleR.html) documentation.

# Software packages
```{r software-load, results="hide", message=FALSE}
library(Seurat)
library(tidyverse)
library(patchwork)
library(pheatmap)
library(RColorBrewer)
library(SingleR)
library(celldex) # To install: BiocManager::install("celldex")
library(data.table)
library(knitr)
library(clusterProfiler)
```

# Data
## Working directory
```{r setwd}
setwd("C:/Users/edlarsen/Documents/240828_scRNAseq/Cluster_Annotation")
```

## Filtered, normalized, and clustered Seurat object(s)
```{r data-import}
ln_seurat <- readRDS(file = "../Normalization_and_Clustering/seurat_NormalizedAndClustered_LN.RData")
```

# Differentially Expressed Features
By default, identifies all positive and negative markers of a single cluster compared to all other cells.

```{r FindAllMarkers-LN, message=FALSE, fig.width = 8, fig.height = 6}

ln_seurat <- FindClusters(ln_seurat, resolution = 0.1)
ln_seurat <- RunUMAP(ln_seurat, dims = 1:10)
ln.markers <- FindAllMarkers(ln_seurat, only.pos=TRUE)
ln.markers %>%
  group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 1)

# export
write.csv(ln.markers, file="LN_DEfeaturesByUMAPcluster_res0.1.csv")

# umap
DimPlot(ln_seurat,
        reduction = "umap",
                   label = TRUE,
                   label.size = 6) + 
  plot_annotation(title = "Canine Lymph Node, Resolution: 0.1", theme = theme(plot.title = element_text(hjust = 0.5, size = 20)))
```

## Top cluster markers
```{r, fig.width = 18, fig.height = 8}
key.genes <- ln.markers[!grepl("ENSCAFG", row.names(ln.markers)),]
key.genes.sortedByPval <- key.genes[order(key.genes$p_val),]
features <- key.genes.sortedByPval %>% group_by(cluster) %>% do(head(., n=5))
features <- as.data.frame(features[!duplicated(features$gene),]) # prioritize unique clusters

DotPlot(ln_seurat,
        assay = "RNA",
        features = rev(features$gene),
        scale = TRUE) + 
  geom_point(aes(size=pct.exp), shape = 21, colour="black", stroke=0.5) +
  labs(y = "Cluster") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```

# Feature plots
Some of these markers were derived from a single-cell atlas of circulating canine leukocytes (https://github.com/dyammons/Canine_Leukocyte_scRNA).
```{r, fig.width = 14, fig.height = 12}
FeaturePlot(ln_seurat, 
            reduction = "umap",
            ncol = 3,
            features = c("CD3E", "CD5", "CD4", "FOXP3", "IL2RA", "GATA3", "TBX21", "RORC")) +
  plot_annotation(title = "Canine Lymph Node: Expression of Canonical CD4 T-cell Markers", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```

```{r, fig.height = 8, fig.width = 14}
FeaturePlot(ln_seurat, 
            reduction = "umap",
            features = c("GNLY", "NKG7", "NCAM1", "KLRK1", "CD8A", "CD3E", "CD5"),
            ncol = 3) +
  plot_annotation(title = "Canine Lymph Node: Expression of Canonical CD8 & NK Cell Markers", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```

```{r, fig.height = 8, fig.width = 10}
FeaturePlot(ln_seurat, 
            reduction = "umap",
            features = c("MS4A1", "CD19", "PAX5", "IRF4")) +
  plot_annotation(title = "Canine Lymph Node: Expression of Canonical B-cell Markers", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```

```{r, fig.height = 8, fig.width = 14}
FeaturePlot(ln_seurat, 
            reduction = "umap",
            ncol = 3,
            features = c("LYZ", "FCER1A", "ITGAM", "CD163", "CD68")) +
  plot_annotation(title = "Canine Lymph Node: Expression of Canonical Monocyte/Macrophage/DC Markers", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```

```{r, fig.height = 8, fig.width = 14}
FeaturePlot(ln_seurat, 
            reduction = "umap",
            features = c("FCER1A", "MMP9", "CCR5", "CXCR2", "EPX", "MMP8", "S100A12", "SERPINA1"),
            ncol = 3) +
  plot_annotation(title = "Canine Lymph Node: Expression of Canonical Granulocyte Markers", theme = theme(plot.title = element_text(hjust = 0.5, size = 20))) &
  scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
```
Note: Granulocytes are difficult to capture with single-cell RNA-seq experiments given their relatively low RNA content and relatively high levels of RNAses and other inhibitory compounds, resulting in fewer detectable transcripts.

# Compare single-cell expression profiles with previously annotated reference datasets using SingleR method
SingleR is a computational method using reference transcriptomic datasets of pure cell types for unbiased cell type recognition in single-cell RNA-seq data. The SingleR method assigns labels to cells based on the reference sample with the highest Spearman rank correlations, using only the marker genes between pairs of labels to focus on the relevant differences between cell types. It also performs a fine-tuning step for each cell, where the correlations are recomputed with just the marker genes for the top-scoring labels. This aims to resolve any ambiguity between those labels by removing noise from irrelevant markers for other labels.

*celldex* contains a number of curated reference datasets, mostly assembled from bulk RNA-seq or microarray data of sorted cell types. Here, a  will be used.

## Blueprint and ENCODE Data
Using a reference constructed from Blueprint and ENCODE data (Martens and Stunnenberg 2013 and ENCODE Project Consortium 2012).
```{r BlueprintEncode}
# pull Blueprint and ENCODE reference data from celldex
ref_bpe <- celldex::BlueprintEncodeData()

# See what cell types are included in the reference and subset to include only those relevant to the single-cell data
unique(ref_bpe$label.main)
ref_bpe <- ref_bpe[,grepl("Neutrophils|Macrophages|Monocytes|Erythrocytes|HSC|Endothelial cells|CD4+ T-cells|DC|CD8+ T-cells|Eosinophils|NK cells|B-cells|Fibroblasts", ref_bpe$label.main)]

# extract normalized counts from Seurat object
norm_counts_ln <- LayerData(ln_seurat, assay = "RNA", layer = "data")
norm_counts_ln <- LayerData(ln_seurat, assay = "RNA", layer = "data")

# use singleR function to annotate cells with labels from reference
bpe_ln <- SingleR(test=norm_counts_ln, ref=ref_bpe, labels=ref_bpe$label.main)
bpe_ln <- SingleR(test=norm_counts_ln, ref=ref_bpe, labels=ref_bpe$label.main)
paste("Lymph node:")
table(bpe_ln$labels)
paste("lnus:")
table(bpe_ln$labels)
```

### Inspect quality of predictions
#### Heatmap
Each column in the heatmap represents a cell. Reference labels on y-axis. We expect cells with a high score for T cells and low for the other clusters to have the final label of T cells.
If prediction for a certain cell is ambiguous, it will have high scores (yellow) for many cell identities.
```{r singleR-bpe-heatmap-ln}
plotScoreHeatmap(bpe_ln)
```

#### Delta distribution
Shows the distribution of the gap between the score for the assigned label and the score of the remaining label(s) across cells assigned to each reference label. We want high deltas, as they give us more confidence in the assigned label.

By default, SingleR will always assign a label to every cell, resulting in incorrect label assignment if the cell's true label isn't in the reference data set. To mitigate this, a "delta" value is calculated for each cell, representing the difference between the score for the assigned label and the median score across all labels. If the delta is small, this indicates that the cell matches all labels with the same confidence, so the assigned label is not very meaningful. SingleR can then discard cells with low delta values caused by ambiguous assignments with closely related reference labels and incorrect assignments that match poorly to all reference labels. 

The 'cleaner' results are found in the pruned_labels column, cells that have deltas that were too small are assigned an NA value:
```{r pruned-cells-bpe}
# The number of cells assigned NA values (deltas too small) will be tallied in the TRUE column
paste("Pruned cells - Lymph node:")
summary(is.na(bpe_ln$pruned.labels))
```

```{r bpe-ln-delta-plot, fig.height = 10, fig.width = 8}
plotDeltaDistribution(bpe_ln, ncol = 4, dots.on.top = FALSE) + ggtitle("Canine Lymph Node Delta Values for Blueprint and ENCODE Annotations")
```

## Human Primary Cell Atlas Data
```{r HPCAtlas}
# pull reference data from celldex
ref_hpca <- celldex::HumanPrimaryCellAtlasData()

# See what cell types are included in the reference and subset to include only those relevant to the single-cell data
unique(ref_hpca$label.main)
ref_hpca <- ref_hpca[,grepl("DC|BM & Prog.|Embryonic_stem_cells|Endothelial_cells|Smooth_muscle_cells|MSC|Pro-Myelocyte|BM|HSC_CD34+|HSC_-G-CSF|GMP|CMP|MEP|Pro-B_cell_CD34-|Pro-B_cell_CD34+|B_cell|Neutrophils|T_cells|Monocyte|Erythroblast|Macrophage|NK_cell|Platelets|Myelocyte", ref_hpca$label.main)]

# use singleR function to annotate cells with labels from reference
hpca_ln <- SingleR(test=norm_counts_ln, ref=ref_hpca, labels=ref_hpca$label.main)
paste("Lymph node:")
table(hpca_ln$labels)
```

### Inspect quality of predictions
#### Heatmap
```{r singleR-hpca-heatmap-ln}
plotScoreHeatmap(hpca_ln)
```

#### Delta distribution
```{r pruned-cells-hpca}
# The number of cells assigned NA values (deltas too small) will be tallied in the TRUE column
paste("Pruned cells - Lymph node:")
summary(is.na(hpca_ln$pruned.labels))
```

```{r hpca-ln-delta-plot, fig.height = 10, fig.width = 8}
plotDeltaDistribution(hpca_ln, ncol = 4, dots.on.top = FALSE) + ggtitle("Canine Lymph Node Delta Values for Human Primary Cell Atlas Annotations")
```

### Using fine labels
```{r HPCAtlas-FineLabels}
# pull reference data from celldex
ref_hpca_fine <- celldex::HumanPrimaryCellAtlasData()

# extract normalized counts from Seurat object
norm_counts_ln <- LayerData(ln_seurat, assay = "RNA", layer = "data")

# use singleR function to annotate cells with labels from reference
hpca_ln_fine <- SingleR(test=norm_counts_ln, ref=ref_hpca_fine, labels=ref_hpca_fine$label.fine)
paste("Lymph node:")
table(hpca_ln_fine$labels)
```

### Inspect quality of predictions
#### Heatmap
```{r singleR-hpcafine-heatmap-ln, fig.width = 16}
plotScoreHeatmap(hpca_ln_fine)
```

#### Delta distribution
```{r pruned-cells-hpcafine}
# The number of cells assigned NA values (deltas too small) will be tallied in the TRUE column
paste("Pruned cells - Lymph node:")
summary(is.na(hpca_ln_fine$pruned.labels))
```

```{r hpcafine-ln-delta-plot, fig.height = 20, fig.width = 14}
plotDeltaDistribution(hpca_ln_fine, ncol = 6, dots.on.top = FALSE) + ggtitle("Canine Lymph Node Delta Values for Human Primary Cell Atlas Annotations")
```

# Cell-Cycle Scoring
See if cluster differences are associated with differences in cell proliferation.

## Assign Cell-Cycle Scores
Each cell will be assigned a score based on its expression of G2/M and S phase markers. The CellCycleStoring() function stores S and G2/M scores in object metadata, along with the predicted classification of each cell in either G2/M, S, or G1 phase.
```{r cell-cycle}
# segregate G2/M phase and S phase markers from Seurat's built-in list of cell cycle markers
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

# assign each cell a score based on expression of G2/M and S phase markers
ln_seurat_cycle <- CellCycleScoring(ln_seurat,
                                s.features = s.genes, 
                                g2m.features = g2m.genes, 
                                set.ident = TRUE)
```

## Visualize cycling cells on UMAP
```{r}
ln_seurat_cycle <- RunPCA(ln_seurat_cycle, features = c(s.genes, g2m.genes))
DimPlot(ln_seurat_cycle,
        group.by = "Phase",
        reduction = "umap",
        pt.size=0.5) + 
  ggtitle("Cell Cycle Scoring: Canine ln")
```

# Module Scoring for Conde et al. human immune cell atlas
Module scoring the average expression levels of each cluster on a single-cell level, subtracted by the aggregated expression of control feature sets. All analyzed features are binned based on averaged expression, and the control features are randomly selected from each bin.
Source: https://cells.ucsc.edu/?ds=pan-immune+global. The top 51 markers were exported from UCSC Cell Browser for each annotated cell type.

```{r, warning = FALSE, message = FALSE, fig.width = 14, fig.height = 10}
# make gene lists
human_ln_atlas_top51 <- read.csv("CondeEtAl_HumanImmuneCellAtlas_Global.csv")
colnames(human_ln_atlas_top51)

plot_list <- list() # initiate empty list for individual plots

for (column in colnames(human_ln_atlas_top51)){
  genes <- list(human_ln_atlas_top51[[column]])
  name <- paste(column, "Score", sep="_")
  ln_seurat <- AddModuleScore(ln_seurat, features = genes, name = name)
  newname <- paste(name, "1", sep="")
  modscoreplot <- FeaturePlot(ln_seurat, features = newname) + scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
  plot_list[[name]] <- modscoreplot # add plot to list
}

# split plot_list into groups of 10 and combine each group of 10 plots into a single layout
plot_groups <- split(plot_list, ceiling(seq_along(plot_list) / 9))
combined_plots <- lapply(plot_groups, function(group) {
  wrap_plots(group, ncol = 3)
})

# display each combined plot
for (i in seq_along(combined_plots)) {
  print(combined_plots[[i]])
}
```

# Module scoring for Ammons canine circulating leukocyte atlas
Source: https://cells.ucsc.edu/?ds=canine-leukocyte-atlas+healthy 

```{r, warning = FALSE, message = FALSE, fig.width = 16, fig.height = 10}
# make gene lists
ammonsk9_ln_atlas_top51 <- read.csv("Ammons_CanineCirculatingLeukocyteAtlas.csv")
colnames(ammonsk9_ln_atlas_top51)

plot_list <- list() # initiate empty list for individual plots

for (column in colnames(ammonsk9_ln_atlas_top51)){
  genes <- list(ammonsk9_ln_atlas_top51[[column]])
  name <- paste(column, "Score", sep="_")
  ln_seurat <- AddModuleScore(ln_seurat, features = genes, name = name)
  newname <- paste(name, "1", sep="")
  modscoreplot <- FeaturePlot(ln_seurat, features = newname) + scale_color_gradientn(colours = brewer.pal(name = "RdPu", n=11))
  plot_list[[name]] <- modscoreplot # add plot to list
}

# split plot_list into groups of 10 and combine each group of 10 plots into a single layout
plot_groups <- split(plot_list, ceiling(seq_along(plot_list) / 9))
combined_plots <- lapply(plot_groups, function(group) {
  wrap_plots(group, ncol = 3)
})

# display each combined plot
for (i in seq_along(combined_plots)) {
  print(combined_plots[[i]])
}
```


# Assigning cell type identity to clusters
To be done once confident cluster assignment has been achieved.
```{r clustID-ln}
new.ln.cluster.ids <- c("LN0_Tcell", "LN1_BCell", "LN2_Tcell", "LN3_Tcell", "LN4_CD8_NK", "LN5_BCell", "LN6_Cycling", "LN7_Myeloid", "LN8_PlasmaCell")
names(new.ln.cluster.ids) <- levels(ln_seurat)
ln_seurat <- RenameIdents(ln_seurat, new.ln.cluster.ids)
DimPlot(ln_seurat, reduction = "umap", label = TRUE, pt.size = 0.5) + ggtitle("Canine Lymph Node, Resolution: 0.1")
saveRDS(ln_seurat, file="LN_Annotated.RData")
```

# Citations
```{r citations}
sessionInfo()
citation()
```